pkg/controller/cluster/cluster_controller.go

```diff
@@ -356,7 +366,96 @@ func (c *Controller) processNextWorkItem() bool {
 	return true
 }
 
+func (c *Controller) buildRemoteClusterClient(cluster *clusteroperator.Cluster) (clusterapiclient.Interface, error) {
+
+	secretName := cluster.Name + "-kubeconfig"
+
+	secret, err := c.kubeClient.CoreV1().Secrets(cluster.Namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	secretData := secret.Data["kubeconfig"]
+
+	config, err := clientcmd.Load(secretData)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{})
+	restConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	clusterAPIClient, err := clusterapiclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	return clusterAPIClient, nil
+}
+
+func (c *Controller) setDeprovisionedComputeMachinesets(cluster *clusteroperator.Cluster) error {
+	original := cluster
+	cluster = cluster.DeepCopy()
+	cluster.Status.DeprovisionedComputeMachinesets = true
+	return controller.PatchClusterStatus(c.client, original, cluster)
+}
+
+func (c *Controller) ensureRemoteMachineSetsAreDeleted(cluster *clusteroperator.Cluster) error {
+	clusterAPIClient, err := c.buildRemoteClusterClient(cluster)
+	if err != nil {
+		return err
+	}
+	machineSets, err := clusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	// If all remote machinesets are gone, then we should update the cluster status
+	if len(machineSets.Items) == 0 {
+		return c.setDeprovisionedComputeMachinesets(cluster)
+	}
+
+	if err = clusterAPIClient.ClusterV1alpha1().MachineSets(remoteClusterNamespace).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{}); err != nil {
+		return err
+	}
+
+	// Ensure that the cluster is enqueued again so we can continue to monitor machinesets
+	c.enqueueAfter(cluster, deleteRemoteInterval)
+	return nil
+}
+
+func (c *Controller) shouldDeleteRemoteMachineSets(cluster *clusteroperator.Cluster) (bool, error) {
+	if cluster.Status.DeprovisionedComputeMachinesets {
+		return false, nil
+	}
+	if len(cluster.Status.MasterMachineSetName) == 0 {
+		return false, nil
+	}
+	masterMachineSet, err := c.machineSetsLister.MachineSets(cluster.Namespace).Get(cluster.Status.MasterMachineSetName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
```

Comment:
Are you sure that we want to swallow this error? The Cluster status indicates that there is a master machine set. I would think that it would be an error worth bubbling up if it turns out that said master machine set does not actually exist.

Reply:
So the problem is that let's say that for some reason the master machine set is deleted, then the cluster itself is marked for deletion. The cluster reconcile loop will not recreate the master machineset and if I return an error from here, we will keep retrying to find a master machineset that doesn't exist.

Reply:
OK. Makes sense.